### ISSUE TITLE
Convert pre-2015 archives to new event format

### ISSUE THREAD

**Issue by igrigorik** [igrigorik/githubarchive.org#112]
Pre-2015 we flattened the JSON events into a pre-defined schema and appended them to a global "timeline" table. This worked, but it has a few issues:

- The GH event format changed overtime, whereas our schema stayed the same
 - New fields were omitted to keep backwards-compat with older data
- One big table was convenient to query but became expensive due to large dataset

Post 2015 we moved to a different format that keeps the minimum number of well-defined fields as separate columns and stashes the rest into a JSON `payload` - see https://www.githubarchive.org/#bigquery for more details.

Ideally, we should reprocess the pre 2015 events to follow the same pattern. The ~rough work involved:

1. Downloading the raw pre-2015 hourly gzip's 
1. Reformat each archive to use the new event format with fixed fields and JSON payload
1. Import above data into separate daily tables
1. Create appropriate rollups to simplify querying 

In other words, an ETL.. with bulk of the work in (2) where all the data massaging needs to happen -- the caveat is that the format changed over time, so this needs to be done carefully to avoid losing data.

_Update: we should also scrub email's from the new tables and files: user@site.com -> redacted@site.com_

**Comment by ronaknnathani** [comment]
Thanks for the description Ilya! It makes things clearer.

I was thinking to start in reverse chronological order from 2014 to 2011 and act on schema based on the changes encountered. If you can share some details regarding how the data is imported in BigQuery tables, it would help in choosing the right tools and formats. I had this idea on the top of my head to use a schema serializer like Avro (https://avro.apache.org/) but I am not sure if that would be the best approach in regards to importing data into the BigQuery tables finally. 

Please let me know what you think and I would be glad to know if you have a specific approach to move forward with.

**Comment by igrigorik** [comment]
The schema we want is here: https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/schema.js - the bulk of the data is in `payload` field. This corresponds to the GitHub format: https://developer.github.com/v3/activity/events/

You can see my current transform logic here: https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/upload.rb#L52

The final format needs to be newline delimited JSON.. same format as above script generates. 

If you're willing to generate the archives and you can upload them to some GCS bucket.. I can kick off the import into BQ.

**Comment by ronaknnathani** [comment]
Thanks for the references! I can generate the archives with newline delimited JSON and the required schema. I will let you know once they are ready to be uploaded.

**Comment by igrigorik** [comment]
@ronaknnathani that would be awesome! FWIW, incremental progress is good too -- we don't need to block on all the archives.

**Comment by ronaknnathani** [comment]
@igrigorik sure, I will move incrementally and keep you posted.

**Comment by harishvc** [comment]
:thumbsup:

**Comment by hsborges** [comment]
:+1:

**Comment by arfon** [comment]
@ronaknnathani - how are you getting on with this? I'd love to help out with this too - I have a copy of the event data pre-2015 and was planning on looking at writing the logic to do this transformation next.

> In other words, an ETL.. with bulk of the work in (2) where all the data massaging needs to happen -- the caveat is that the format changed over time, so this needs to be done carefully to avoid losing data.

I realize it's somewhat embarrassing to be asking this given my role/current employer 😊 but do you know how often this happened?

**Comment by arfon** [comment]
> @igrigorik I realize it's somewhat embarrassing to be asking this given my role/current employer 😊 but do you know how often this happened?

Further to my last post, here are some of the changes I've stumbled upon during preliminary processing:

The `PushEvent` `payload` used to have a `shas` array containing the commit information (including the committer). This key later changed to `commits`.

Event `display_login` (https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/upload.rb#L54) looks to be a relatively new thing. Older events only seem to have a `login` field.

@igrigorik - if you have a chance, would appreciated your feedback on the following:

- Sometimes events don't parse and this looks to be a GitHub thing: https://github.com/brianmario/yajl-ruby/issues/64#issuecomment-4440034 - did you end up just dropping those from the archive?
- What do we do when the actor is `nil`, for example when there are anonymous Gists: https://help.github.com/articles/creating-gists/#creating-an-anonymous-gist - do we just record these events with a [blank `actor` field](https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/schema.js#L40)?

**Comment by igrigorik** [comment]
@arfon thanks for working on this! Before I get into the specifics, some high-level points:

1. My first attempt at this whole problem was to [auto-generate a "flattened" schema](https://gist.github.com/igrigorik/83334277835625916cd6) for BigQuery. (Note: at the time, BigQuery did not support nested records, hence the flatten logic).
 - Pros: all fields are mapped to individual columns, which means you can do targeted (fast and cheap) selects, filters, etc.
 - Cons: we're locked into the schema.. If anything changes, we either have start a new table with updated schema, reimport all old data with new field as "null", or drop the new field and stick with old schema. 
  - We ran with "drop new field" for a while.. with obvious downsides.

2. Second attempt, which is what we're doing now, takes a hybrid approach:
 - Extract "stable fields" into a predefined schema
 - Serialize payloads that are likely to change into JSON string, that can be parsed via JSON_EXTRACT..
 - Pro: you can still do fast and cheap selects on most popular fields, new fields can be added and are available immediately (no changes needed in the pipeline). Assuming "stable" fields are, actually stable :)
 - Cons: some fields are inside of the JSON blob, which requires an expensive parse. 

-- 

We ran with 2 for ~19 months before [display_login showed up](https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/upload.rb#L54). So, on balance it seems to have worked pretty well.. However, now that I think about it, I think we can improve things a bit:
 1. We can freeze the top-level schema ([same as now](https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/schema.js))
 2. As part of the upload pipeline, we can validate the top-level fields and find/detect any that are not in the schema --> instead of dropping them, we can serialize them into a JSON object.. e.g `"other": "json_string here"`. This way we never drop data. 

(2) is a relatively simple fix on top of what we have currently.

--

Back to import pre 2015 data...
- Ideally, massage it into [same top-level schema](https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/schema.js)
 - Extract actor / repo / org / created_at / type / id
 - Push everything else into payload (or, we could even push the _full_ record into it, including above fields)

> Sometimes events don't parse and this looks to be a GitHub thing: brianmario/yajl-ruby#64 (comment) - did you end up just dropping those from the archive?

Hmm, I was under the impression that we resolved this (by fixing Yajl). Not the case?

> What do we do when the actor is nil, for example when there are anonymous Gists: https://help.github.com/articles/creating-gists/#creating-an-anonymous-gist - do we just record these events with a blank actor field?

Yep. And that also reminds me... https://github.com/igrigorik/githubarchive.org/issues/123

**Comment by arfon** [comment]
@igrigorik - thanks for the additional context ⚡

> We can freeze the top-level schema (same as now)

👍 

> As part of the upload pipeline, we can validate the top-level fields and find/detect any that are not in the schema --> instead of dropping them, we can serialize them into a JSON object.. e.g `"other": "json_string here"`. This way we never drop data.

Sounds good. I'm thinking we could do with a small Ruby module that can be included in https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/upload.rb that transforms the events into the schema (and adds any extra fields to the `other` field). This module should be able to handle _all_ events (past and current) incase there's a reason to process all the events again sometime.

I've half-written this already while testing the parsing of these archival events so if it sounds like a good approach I'll keep pushing on that.

> We ran with 2 for ~19 months before display_login showed up. So, on balance it seems to have worked pretty well.. However, now that I think about it, I think we can improve things a bit:

:smile: I hadn't realized that was such a recent addition. BTW I've dug into this a little in the GitHub codebase and it looks like we are safe to drop from the `actor` key for now (but perhaps put in an `other` field as you suggest).

> Hmm, I was under the impression that we resolved this (by fixing Yajl). Not the case?

I think it was fixed. It's just there seem to be some events in the tar files that aren't parseable by Yajl (it's pretty small numbers).

**Comment by igrigorik** [comment]
> This module should be able to handle all events (past and current) incase there's a reason to process all the events again sometime... I've half-written this already while testing the parsing of these archival events so if it sounds like a good approach I'll keep pushing on that.

Awesome!

> BTW I've dug into this a little in the GitHub codebase and it looks like we are safe to drop from the actor key for now (but perhaps put in an other field as you suggest).

Two options to explore here:
- (a) We can craft the the top-level ("frozen") fields and whatever is the remainder we serialize and stick into top-level "other" JSON-String field -- e.g. say tomorrow you guys add a new key under 'repo' and 'actor'. Our top level would stay the same, but then in 'other' we'd have --> "{repo: {'newthing': ...}, 'actor': {'othernewthing': ...}}".
- (b) or, we could put "other" under each object - e.g. actor record has 'other', record does too...

Personally, I think I'm leaning towards (a) -- simpler to grok and implement, I think.

Actually, nm, .. I think we _want_ (a) because then we only need to update our top-level schema once with one extra field, instead of defining "other" on each record object.

Does this make sense?

**Comment by arfon** [comment]
> Does this make sense?

Yep, that makes complete sense. BTW, here's the direction I've been heading with the event parser: https://gist.github.com/arfon/53c570d157e5b1ce166d68a87912e0ab . 

> (a) We can craft the the top-level ("frozen") fields and whatever is the remainder we serialize and stick into top-level "other" JSON-String field -- e.g. say tomorrow you guys add a new key under 'repo' and 'actor'. Our top level would stay the same, but then in 'other' we'd have --> "{repo: {'newthing': ...}, 'actor': {'othernewthing': ...}}".

👍 This is basically what the `parse_field` method does [here](https://gist.github.com/arfon/53c570d157e5b1ce166d68a87912e0ab#file-event_transform-rb-L60-L82). 

BTW, the only major structural difference in events between 2012-2016 I'm handling is in the structure of the payload in the `PushEvent` (when scrubbing emails) which happens [here](https://gist.github.com/arfon/53c570d157e5b1ce166d68a87912e0ab#file-event_transform-rb-L89-L116). Does that sound about right?

If this seems like a good direction I'll write some tests based using some real events as fixtures.

**Comment by igrigorik** [comment]
> https://gist.github.com/arfon/53c570d157e5b1ce166d68a87912e0ab .

That's looking great! Have a feeling it'll get a bit more complex once we start looking at older records though.. which had different format and such?

> This is basically what the parse_field method does here.

Other should probably preserve the nested structure? Right now I think it'll just hoist any and all missing subfields into same top-level namespace (which can conflict too). Instead, I think we want to preserve the nested structure?


> BTW, the only major structural difference in events between 2012-2016 I'm handling is in the structure of the payload in the PushEvent (when scrubbing emails) which happens here. Does that sound about right?

Ah, interesting. Are you finding that all the other fields match?

**Comment by arfon** [comment]
> Other should probably preserve the nested structure? Right now I think it'll just hoist any and all missing subfields into same top-level namespace (which can conflict too). Instead, I think we want to preserve the nested structure?

So as currently implemented, `"actor": { "display_login" : "arfon" }` would end up sitting in `"other" : {"actor" : { "display_login" : "arfon" }}`. Is that sufficient or do you think we need to keep track of the full path of the additional fields?

> Ah, interesting. Are you finding that all the other fields match?

I *think* so. This is why I keep asking you, I'd expected the format to have been more varied :-). I'm going to sanity check this by grabbing a few different event types across all the archive files and comparing their structure.

**Comment by igrigorik** [comment]
> So as currently implemented, "actor": { "display_login" : "arfon" } would end up sitting in "other" : {"actor" : { "display_login" : "arfon" }}. 

That's what I was shooting for as well. 👍

**Comment by arfon** [comment]
> That's looking great! Have a feeling it'll get a bit more complex once we start looking at older records though.. which had different format and such?

@igrigorik - looks like I had a rogue `rescue NoMethodError` in my parsing script which was silently catch ing a bunch of failures. The format has indeed changed a fair bit over time.

**Comment by arfon** [comment]
@igrigorik - so inspecting a few different `PushEvent` structures across the years it looks like there's some variability.

In particular it looks like the [2013-era](https://gist.github.com/arfon/c12da8cf8efdf65b387765a4a8813078#2013-02-15-0jsongz) events have `actor` and `actor_attributes` keys that were later replaced by just the `actor`key.

Also, some 2013/2014 events don't have the full org information. See [this one](https://gist.github.com/arfon/c12da8cf8efdf65b387765a4a8813078#2014-02-15-0jsongz) for example. There's an organization field on `repository` (`"organization":"slic"`) but that's the only field we have.

This raises the question, give that our org schema looks like [this](https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/schema.js#L72-L101) we can sometimes only populate the `login` field. I suspect there's not much we can do about this but wanted to check with you just incase you had any ideas.

**Comment by igrigorik** [comment]
> This raises the question, give that our org schema looks like this we can sometimes only populate the login field. I suspect there's not much we can do about this but wanted to check with you just incase you had any ideas.

Yeah, I don't think we want to go any further than massaging provided data into common fields: if data is missing, it'll be a blank field; unrecognized fields go into "other"; payload is serialized as a string.

**Comment by arfon** [comment]
@igrigorik - I've done a bunch more on this today and at this point could do with a review from yourself 🙏 

I've created a [repo](https://github.com/arfon/gh_archive_parser) with the [event parser](https://github.com/arfon/gh_archive_parser/blob/master/event_transform.rb), some [test files from GitHub Archive](https://github.com/arfon/gh_archive_parser/tree/master/files) and a [bunch of fixtures](https://github.com/arfon/gh_archive_parser/tree/master/fixtures) from across the years (2011--2016) so that as we modify the event parser we can be sure we're not breaking the schema. 

**What I could do with a sanity check on now...**

For each year (e.g. [2011](https://github.com/arfon/gh_archive_parser/tree/master/fixtures/2011)) there is a folder with three files for each event type (e.g. `CreateEvent`):

- `CreateEvent_raw.json` the raw JSON from the file downloaded from `http://data.githubarchive.org`
- `CreateEvent_transformed.json` the _transformed_ JSON (i.e. after being parsed by `EventTransform`)
- `CreateEvent_encoded.json` the _encoded_ JSON (i.e. coming from [here](https://github.com/arfon/gh_archive_parser/blob/master/parse.rb#L26))

I've pretty-printed the `raw` and `transformed` JSON files so they're easier to inspect but I'd appreciate a sanity check of the structure of these files.

In addition, any/all emails in the original `_raw` files should now be scrubbed.

Finally, there are some tests for checking that all transformed files have the [correct top-level structure](https://github.com/arfon/gh_archive_parser/blob/master/parse.rb#L26) and also a test for the [`EventTransform` class](https://github.com/arfon/gh_archive_parser/blob/master/event_transform_test.rb#L40-L53) that compares the output of `EventTransform` against the `_transformed` fixture files (which are currently considered the 'correct' output).

**Some general notes after staring at this for too long**

- Broadly speaking the top-level event format in 2011/2012/2015/2016 are very similar.
- 2013/2014 are consistent with each other (but different from ☝). I believe this is when GitHub Archive was pulling from the (now deprecated) `timeline.json`
- Events from `timeline.json` don't seem to have the `id` field set (e.g. https://github.com/arfon/gh_archive_parser/blob/master/fixtures/2014/PushEvent_raw.json)
- Events from `timeline.json` also don't have an `org` field but if the repository is owner by and organization (e.g. [here](https://github.com/arfon/gh_archive_parser/blob/master/fixtures/2014/TeamAddEvent_raw.json#L131)) then we extract that field into the [transformed JSON](https://github.com/arfon/gh_archive_parser/blob/master/fixtures/2014/TeamAddEvent_transformed.json#L11-L13).

That's probably enough for now!

**Comment by igrigorik** [comment]
@arfon apologies about the delay! 🙇 

This is looking really good. A few comments as I'm scanning through the examples:

- BigQuery [doesn't like `null` values in JSON](http://stackoverflow.com/questions/32617550/bigquery-loading-json-fields-with-null-values)—grr. Current transform logic sets missing keys to `null`, which will cause trouble at import time. We should probably drop null's.
- 2011 events have [funky "repo" values](https://github.com/arfon/gh_archive_parser/blob/master/fixtures/2011/DeleteEvent_transformed.json#L18): api.github.dev.. weird! Should we keep those?
- Aside: event ID's are strings.. but that's how GitHub API returns them, I guess we'll keep those as is. *shrug*

I'm guessing next step here would be to transform some subset of files from each year and actually try to run the BQ imports.. see if we run into any issues that we haven't spotted yet.

**Comment by arfon** [comment]
Thanks for the feedback @igrigorik. 

> - BigQuery doesn't like null values in JSON—grr. Current transform logic sets missing keys to `null`, which will cause trouble at import time. We should probably drop null's.

OK, I've fixed that in https://github.com/arfon/gh_archive_parser/commit/1a25b579786f468171fb448a61d068c2573f0111 . If you could take a look at some of the new `_transformed.json` files [here](https://github.com/arfon/gh_archive_parser/tree/master/fixtures)?

> - 2011 events have funky "repo" values: api.github.dev.. weird! Should we keep those?

Lol, yeah I saw those. I think we should keep them as-is.

> - Aside: event ID's are strings.. but that's how GitHub API returns them, I guess we'll keep those as is. shrug

OK, I've left them as strings.

> I'm guessing next step here would be to transform some subset of files from each year and actually try to run the BQ imports.. see if we run into any issues that we haven't spotted yet.

Sounds good. I've transformed a bunch of the events and added them here https://github.com/arfon/gh_archive_parser/tree/master/files - see files with `transformed` in the filename (e.g. `2011-02-15-0.transformed.json.gz`). Would you be able to try these out?

**Comment by igrigorik** [comment]
> OK, I've fixed that in arfon/gh_archive_parser@1a25b57 . If you could take a look at some of the new _transformed.json files here?

Nice! One more thing, we should serialize "other" in the same way as payload: https://github.com/arfon/gh_archive_parser/blob/1a25b579786f468171fb448a61d068c2573f0111/fixtures/2014/PushEvent_transformed.json#L14 -- in BQ we'll treat it as a STRING field.

**Comment by arfon** [comment]
> Nice! One more thing, we should serialize "other" in the same way as payload: https://github.com/arfon/gh_archive_parser/blob/1a25b579786f468171fb448a61d068c2573f0111/fixtures/2014/PushEvent_transformed.json#L14 -- in BQ we'll treat it as a STRING field.

Will do. Just to check, that can also be missing if there's nothing in there (rather than being `NULL`)?

**Comment by arfon** [comment]
@igrigorik - ok, `other` is encoded as a string now too: https://github.com/arfon/gh_archive_parser/commit/dbea1999f8d8d2ca6f5c1e444998c42cc0a5ccd6

**Comment by igrigorik** [comment]
✨ 🎉 ✨ 🎉 ✨ 🎉 ✨ 🎉  Done! ✨ 🎉 ✨ 🎉 ✨ 🎉 ✨ 🎉 ✨ 

- All gzip archives have been reprocessed  and updated (same location as before)
- **"daily" dataset now contains the full history (2011-now 🎉 ) for each day**
    - The top-level [schema is the same](https://github.com/igrigorik/githubarchive.org/blob/master/bigquery/schema.js) for all days
    - "payload" field contains the JSON-encoded payload, if provided by the event.
    - "other" field contains the JSON-encoded content of any other fields not captured by the schema.
    - **new location**: `day.events_YYYMMMDD` --> `day.YYYMMMDD` -- the `_events` prefix is no longer necessary. Update your queries and remove the prefix.
- "monthly" and "yearly" datasets have been reprocessed on top of the new "daily" data
- githubarchive.org landing page is updated with new queries and information about the schema.

Huge thanks to @arfon for doing the heavy lifting on reprocessing!

**Comment by arfon** [comment]
🎉  ⚡ 🚀
